lib/vrl/compiler/src/expression/abort.rs

```diff
@@ -31,3 +79,80 @@ impl fmt::Display for Abort {
         write!(f, "abort")
     }
 }
+
+// -----------------------------------------------------------------------------
+
+#[derive(Debug)]
+pub struct Error {
+    variant: ErrorVariant,
+    expr_span: Span,
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum ErrorVariant {
+    #[error("non-string abort message")]
+    NonString(Kind),
+    #[error("unhandled fallible expression")]
+    FallibleExpr,
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{:#}", self.variant)
+    }
+}
+
+impl std::error::Error for Error {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        Some(&self.variant)
+    }
+}
+
+impl DiagnosticError for Error {
+    fn code(&self) -> usize {
+        use ErrorVariant::*;
+
+        match self.variant {
+            NonString(_) => 300,
+            FallibleExpr => 630,
```

Comment:
The first one can be removed if we use `value::Error::Expected`, the second one would have to be documented in our errors page (check our `cue` docs).

Reply:
Hmm, yeah I think we should create a new one, given that that one explicitly talks about function arguments, which this is not. It could potentially confuse users into thinking `abort` is a function call, and they can treat it as such, which they can not.

Reply:
What is the protocol for creating a new error code? Not sure if I missed a README, but I can't quite tell what the pattern is for a new code. Would I make it `631`?

Reply:
Answered this over chat with Jean -- we'll proceed with `631` as the new error code. In general, error codes will likely be overhauled with a more formal pattern/process in the future.

Reply:
Added in 678d5aa